program/rust/src/processor/upd_price.rs

```diff
@@ -173,25 +216,42 @@ pub fn upd_price(
                 clock.slot,
                 clock.unix_timestamp,
             )
-        };
-
-        // If the aggregate was successfully updated, calculate the difference and update TWAP.
-        if updated {
-            let agg_diff = (clock.slot as i64)
-                - load_checked::<PriceAccount>(price_account, cmd_args.header.version)?.prev_slot_
-                    as i64;
-            // Encapsulate TWAP update logic in a function to minimize unsafe block scope.
-            unsafe {
-                c_upd_twap(price_account.try_borrow_mut_data()?.as_mut_ptr(), agg_diff);
-            }
+        }
+    };
+
+    // If the aggregate was successfully updated, calculate the difference and update TWAP.
+    if updated {
+        let agg_diff = (clock.slot as i64)
+            - load_checked::<PriceAccount>(price_account, cmd_args.header.version)?.prev_slot_
+                as i64;
+        {
             let mut price_data =
                 load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
+
+            if !(noninitial_price_update_after_program_upgrade) {
+                // Multiple price updates may occur within the same slot. Updates within the same slot will
+                // use the previously calculated values (prev_twap, prev_twac, and prev_price_cumulative)
+                // from the last successful aggregated price update as their basis for recalculation. This
+                // ensures that each update within a slot builds upon the last and not the twap/twac/price_cumulative
+                // that is calculated right after the publishers' individual price updates.
+                if slots_since_last_update > 0 {
+                    price_data.prev_twap_ = price_data.twap_;
+                    price_data.prev_twac_ = price_data.twac_;
+                    price_data.prev_price_cumulative = price_data.price_cumulative;
+                }
+                price_data.twap_ = price_data.prev_twap_;
+                price_data.twac_ = price_data.prev_twac_;
+                price_data.price_cumulative = price_data.prev_price_cumulative;
+            }
+            price_data.update_price_cumulative()?;
             // We want to send a message every time the aggregate price updates. However, during the migration,
             // not every publisher will necessarily provide the accumulator accounts. The message_sent_ flag
             // ensures that after every aggregate update, the next publisher who provides the accumulator accounts
             // will send the message.
             price_data.message_sent_ = 0;
-            price_data.update_price_cumulative()?;
+        }
+        unsafe {
```

Comment:
Can you please move the computation of agg_diff and the setting of the needed twap_ variables to above this line so that the relationship between these pieces of the code is clear to the reader?